subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/bean/AbstractNestedRuntimeBeanNode.java

```diff
@@ -51,14 +50,11 @@ protected AbstractNestedRuntimeBeanNode(@Nullable RuntimeBeanNode<?> parentNode,
 
     public void visitProperties(PropertyVisitor visitor, PropertySpecFactory specFactory, final Queue<RuntimeBeanNode<?>> queue, final RuntimeBeanNodeFactory nodeFactory) {
         TypeMetadata typeMetadata = getTypeMetadata();
-        for (final PropertyMetadata propertyMetadata : typeMetadata.getPropertiesMetadata()) {
-            PropertyValueVisitor propertyValueVisitor = propertyMetadata.getPropertyValueVisitor();
-            if (propertyValueVisitor == null || !propertyValueVisitor.shouldVisit(visitor)) {
-                continue;
-            }
+        for (final WorkPropertyMetadata workMetadata : typeMetadata.getPropertiesMetadata()) {
+            PropertyMetadata propertyMetadata = workMetadata.getPropertyMetadata();
             String propertyName = getQualifiedPropertyName(propertyMetadata.getFieldName());
-            PropertyValue propertyValue = new DefaultPropertyValue(propertyName, propertyMetadata, getBean(), propertyMetadata.getMethod());
-            propertyValueVisitor.visitPropertyValue(propertyValue, visitor, specFactory, new BeanPropertyContext() {
+            PropertyValue propertyValue = new DefaultPropertyValue(propertyName, workMetadata.getPropertyMetadata(), getBean());
+            workMetadata.visitPropertyValue(propertyValue, visitor, specFactory, new BeanPropertyContext() {
```

Comment:
I think most of the changes in this class should be reverted if we merge `WorkPropertyMetadata` and `PropertyMetadata`.